### PR TITLE
Seun save state

### PR DESCRIPTION
**Pull request by ogewan**
Minimum Viable Product

### CODE REVIEW

File: src/neuroglancer/save_state/save_state.ts [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+
+export class SaveState extends RefCounted {
+  private activeKey?: string|null;
+  lastKey?: string|null;
+  saveStorage: any = {};
+  supported = true;
+  constructor(public root: Trackable, updateDelayMilliseconds = 400) {
+    super();
+    const saverManuallyDisabled = !getOldStyleSaving().value;
+    this.supported = !storageAvailable() && saverManuallyDisabled;
```
Comment (nkemnitz) [inline src/neuroglancer/save_state/save_state.ts]: This seems confusing - state saving is supported when there is no storage available and the user manually disabled state saving? Either the check or the variable names don't match...

**Comment by ogewan** [comment]
Ready for Review

**Comment by ogewan** [comment]
Significant changes since last review:
sid changed to local_id
local_id corresponds to a tabbed session and no longer changes
Individual states are no longer saved, it is grouped into the tabbed session
save button changed to share
sharing state provides json link and stores in history if state server is accessible
sharing a state that has not been updated since last share will provide the same links provided previously without pushing to state server.

keyboard shortcuts:
ctrl + shift + s = save and save window
ctrl + shift + j = save and copy json link or raw link if no json link
ctrl + shift + r = save and copy raw link

File: src/neuroglancer/viewer.ts [hunk 2]
```diff
@@ -773,32 +812,69 @@ export class Viewer extends RefCounted implements ViewerState {
...
           .catch(() => {
             this.promptJsonStateServer('state server not responding, enter a new one?');
             if (this.jsonStateServer.value) {
               this.postJsonState();
             }
+          })
+          .finally(() => {
+            const noStateServerAccess = !this.jsonStateServer.value;
+            if (noStateServerAccess) {
+              this.hashBinding!.legacy.setUrlHash();
```
Comment (nkemnitz) [inline src/neuroglancer/viewer.ts]: `hashBinding` _could_ be undefined, right? I thought it's only initialized if local storage does not work?

File: src/neuroglancer/viewer.ts [hunk 3]
```diff
@@ -773,32 +812,69 @@ export class Viewer extends RefCounted implements ViewerState {
...
+              this.showSaveDialog();
+            }
           });
     }
+    else {
+      StatusMessage.showTemporaryMessage(`Cannot access state server. Press the share button/CTRL + SHIFT + S to enter a server URL.`);
+      this.showSaveDialog(2);
```
Comment (nkemnitz) [inline src/neuroglancer/viewer.ts]: Use `UrlType.raw`

File: src/neuroglancer/viewer.ts [hunk 4]
```diff
@@ -202,9 +207,16 @@ function makeViewerContextMenu(viewer: Viewer) {
...
 
+export enum UrlType {
+  json = 1,
+  raw,
+}
+
 export class Viewer extends RefCounted implements ViewerState {
   navigationState = this.registerDisposer(new NavigationState());
   perspectiveNavigationState = new NavigationState(new Pose(this.navigationState.position), 1);
+  saver?: SaveState;
```
[the remark this answers is not in the document]
Comment (ogewan) [reply src/neuroglancer/viewer.ts]: Unfortunately, it depends on loadFromJson which is async and cannot be done in the constructor. But I moved initialization of setup into Viewer

File: src/neuroglancer/viewer.ts [hunk 5]
```diff
@@ -202,9 +207,16 @@ function makeViewerContextMenu(viewer: Viewer) {
...
+  hashBinding?: UrlHashBinding;
```
[the remark this answers is not in the document]
Comment (ogewan) [reply src/neuroglancer/viewer.ts]: Similarly, optional hashbinding depends on saver which cannot be constructed till the async load from json; moved to the same place as saver initialization.

File: src/neuroglancer/viewer.ts [hunk 6]
```diff
@@ -773,32 +812,69 @@ export class Viewer extends RefCounted implements ViewerState {
...
+                delay: true,
+                errorPrefix: `Error retrieving state: `,
+              })
+          .finally(() => {
+            this.resetStateWhenEmpty = true;
+          });
     }
+  }
 
+  postJsonState(get: UrlType = 0) {
```
[the remark this answers is not in the document]
Comment (ogewan) [reply src/neuroglancer/viewer.ts]: UrlType = 0 is equivalent to not setting the UrlType which is the case for CTRL + SHIFT + S and pressing the share button normal.

[the remark this answers is not in the document]
Comment (ogewan) [reply src/neuroglancer/viewer.ts]: changed UrlType to a conditional parameter

File: src/neuroglancer/preferences/user_preferences.ts [hunk 7]
```diff
@@ -10,19 +10,19 @@ class UserPreferences {
   renderMeshByDefault: TrackableBoolean;
   prefetchSliceViewChunks: TrackableBoolean;
   cursorOnMousedrag: TrackableBoolean;
-  urlAutoSave: TrackableBoolean;
+  oldStyleSave: TrackableBoolean;
```
[the remark this answers is not in the document]
Comment (ogewan) [reply src/neuroglancer/preferences/user_preferences.ts]: changed to saveToAddressBar

File: src/neuroglancer/viewer.ts (hunk 2 above)
Comment (ogewan) [reply src/neuroglancer/viewer.ts]: added a comment explaining. hashBinding is not optional hashBinding.legacy.falback is

File: src/neuroglancer/save_state/save_state.ts [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+  constructor(public viewer: Viewer, jsonString?: string, get?: UrlType) {
+    super();
+
+    let urlStart = `${window.location.origin}${window.location.pathname}`;
+    let jsonUrl;
+    if (jsonString) {
+      jsonUrl = `${urlStart}?json_url=${jsonString}`;
+    } else {
+      if (!jsonUrl) {
+        jsonUrl = 'NOT AVALABLE';
```
Comment (nkemnitz) [inline src/neuroglancer/save_state/save_state.ts]: Typo: "available"
Maybe make it a constant to ensure that we are calling it the same everywhere in the code.

File: src/neuroglancer/save_state/save_state.ts [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+    let form = document.createElement('form');
+    let {content} = this;
+    content.style.overflow = 'visible';
+
+    let modal = document.createElement('div');
+    content.appendChild(modal);
+
+    form.append(this.makePopup('JSON_URL'));
+    this.insertField(
+        form, 'JSON_URL', jsonUrl, 'neuroglancer-save-state-json', jsonUrl === 'NOT AVALABLE');
```
Comment (nkemnitz) [inline src/neuroglancer/save_state/save_state.ts]: Typo: "available", see earlier comment.

File: src/neuroglancer/save_state/save_state.ts [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+    new SaveHistoryDialog(viewer, this);
+  }
+}
+
+class SaveDialog extends Overlay {
+  constructor(public viewer: Viewer, jsonString?: string, get?: UrlType) {
+    super();
+
+    let urlStart = `${window.location.origin}${window.location.pathname}`;
+    let jsonUrl;
```
Comment (nkemnitz) [inline src/neuroglancer/save_state/save_state.ts]: Can just initialize `jsonURL` here and remove the `else` branch below.

File: src/neuroglancer/save_state/save_state.ts [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+    new SaveDialog(viewer, jsonString, get);
+  }
+
+  showHistory(viewer: Viewer) {
+    new SaveHistoryDialog(viewer, this);
+  }
+}
+
+class SaveDialog extends Overlay {
+  constructor(public viewer: Viewer, jsonString?: string, get?: UrlType) {
```
Comment (nkemnitz) [inline src/neuroglancer/save_state/save_state.ts]: `get` should be a bit more descriptive, e.g. `getUrlType`

File: src/neuroglancer/viewer.ts [hunk 12]
```diff
@@ -742,27 +790,18 @@ export class Viewer extends RefCounted implements ViewerState {
...
-            errorPrefix: `Error retrieving state: `,
-          });
-    }
-  }
-
   showReportDialog(image: string) {
     new UserReportDialog(this, image);
   }
 
+  showSaveDialog(get?: UrlType, jsonString?: string) {
```
Comment (nkemnitz) [inline src/neuroglancer/viewer.ts]: Copied from `SaveDialog`: `get` should be a bit more descriptive, e.g. `getUrlType`

File: src/neuroglancer/viewer.ts [hunk 13]
```diff
@@ -773,31 +812,70 @@ export class Viewer extends RefCounted implements ViewerState {
...
+                delay: true,
+                errorPrefix: `Error retrieving state: `,
+              })
+          .finally(() => {
+            this.resetStateWhenEmpty = true;
+          });
     }
+  }
 
+  postJsonState(get?: UrlType) {
```
Comment (nkemnitz) [inline src/neuroglancer/viewer.ts]: Copied from `SaveDialog`: `get` should be a bit more descriptive, e.g. `getUrlType`

**Pull request merged**